src/Aspire.Hosting.Kubernetes/KubernetesResourceContext.cs

```diff
@@ -0,0 +1,368 @@
+    public readonly Dictionary<string, HelmExpressionValue> Secrets = [];
+    public readonly Dictionary<string, HelmExpressionValue> Parameters = [];
+    public Dictionary<string, string> Labels = [];
+    public List<BaseKubernetesResource> TemplatedResources { get; } = [];
+    internal List<string> Commands { get; } = [];
+    internal List<VolumeMountV1> Volumes { get; } = [];
+    internal IResource Resource => resource;
+    internal ILogger Logger => kubernetesPublishingContext.Logger;
+    internal KubernetesPublisherOptions PublisherOptions => publisherOptions;
+
+    public void BuildKubernetesResources()
+    {
+        SetLabels();
+        CreateApplication();
+        AddIfExists(resource.ToConfigMap(this));
+        AddIfExists(resource.ToSecret(this));
+        AddIfExists(resource.ToService(this));
+    }
+
+    private void SetLabels()
+    {
+        Labels = new()
+        {
+            ["app"] = "aspire",
+            ["component"] = resource.Name,
+        };
+    }
+
+    private void CreateApplication()
+    {
+        if (resource is IResourceWithConnectionString)
+        {
+            var statefulSet = resource.ToStatefulSet(this);
+            TemplatedResources.Add(statefulSet);
+            return;
+        }
+
+        var deployment = resource.ToDeployment(this);
+        TemplatedResources.Add(deployment);
+    }
+
+    private void AddIfExists(BaseKubernetesResource? instance)
+    {
+        if (instance is not null)
+        {
+            TemplatedResources.Add(instance);
+        }
+    }
+
+    internal bool TryGetContainerImageName(IResource resourceInstance, out string? containerImageName)
+    {
+        if (!resourceInstance.TryGetLastAnnotation<DockerfileBuildAnnotation>(out _) && resourceInstance is not ProjectResource)
+        {
+            return resourceInstance.TryGetContainerImageName(out containerImageName);
+        }
+
+        var imageEnvName = $"{resourceInstance.Name.ToManifestFriendlyResourceName()}_image";
+        var value = $"{resourceInstance.Name}:latest";
+        var expression = imageEnvName.ToHelmParameterExpression(resource.Name);
+
+        Parameters[imageEnvName] = new HelmExpressionValue(expression, value);
+        containerImageName = expression;
+        return false;
+
+    }
+
+    public async Task ProcessResourceAsync(DistributedApplicationExecutionContext executionContext, CancellationToken cancellationToken)
+    {
+        ProcessEndpoints();
+        ProcessVolumes();
+
+        await ProcessEnvironmentAsync(executionContext, cancellationToken).ConfigureAwait(false);
+        await ProcessArgumentsAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    private void ProcessEndpoints()
+    {
+        if (!resource.TryGetEndpoints(out var endpoints))
+        {
+            return;
+        }
+
+        foreach (var endpoint in endpoints)
+        {
+            var port = endpoint.TargetPort ?? 80;
```

Comment:
We should only default to 80 for http no? Otherwise it should fail?

Reply:
aye we should probably take the scheme into account here

Reply:
Should we default to a port number > 1024?  Ports below 1024 require root permissions or reconfiguring the `net.ipv4.ip_unprivileged_port_start` security context.

Reply:
Good call. Will switch to 8080 for http scheme, 8443 for https and 9000 for everything else